osgar/drivers/tankmaria.py

```diff
@@ -0,0 +1,149 @@
+        self.desired_angular_speed = 0.0
+
+        # status
+        self.emergency_stop = None  # uknown state
+        self.pose = (0.0, 0.0, 0.0)  # x, y in meters, heading in radians (not corrected to 2PI)
+        self.buttons = None
+
+        self.verbose = False  # should be in Node
+        self.buf = b''
+        self.encoder = None
+        self.led = 1
+        self.milisec = None
+
+    def send_pose(self):
+        x, y, heading = self.pose
+        self.publish('pose2d', [round(x*1000), round(y*1000),
+                                round(math.degrees(heading)*100)])
+
+
+    def compute_pose(self, left, right):
+        """Update internal pose with 'dt' step"""
+        if left is None or right is None:
+            return False, None, None
+        x, y, heading = self.pose
+
+        metricL = ENC_SCALE * left  # dt is already part of ENC_SCALE
+        metricR = ENC_SCALE * right
+
+        dist = (metricL + metricR)/2.0
+        angle = (metricR - metricL)/WHEEL_DISTANCE
+
+        # advance robot by given distance and angle
+        if abs(angle) < 0.0000001:  # EPS
+            # Straight movement - a special case
+            x += dist * math.cos(heading)
+            y += dist * math.sin(heading)
+            #Not needed: heading += angle
+        else:
+            # Arc
+            r = dist / angle
+            x += -r * math.sin(heading) + r * math.sin(heading + angle)
+            y += +r * math.cos(heading) - r * math.cos(heading + angle)
+            heading += angle # not normalized
+        return True, (x, y, heading), (dist, angle)
+
+    def get_packet(self, buf):
+        s = buf.split(b'\n')
+        if len(s) < 2:
+            return None, buf
+        return s[0].strip(), b'\n'.join(s[1:])
+
+    def slot_desired_speed(self, timestamp, data):
+        self.desired_speed, self.desired_angular_speed = data[0]/1000.0, math.radians(data[1]/100.0)
+
+    def slot_raw(self, timestamp, data):
+        self.buf += data
+        packet, self.buf = self.get_packet(self.buf)
+        while packet is not None:
+            try:
+                d = [int(x) for x in packet.split()]
+            except ValueError:
+                print('Invalid packet', packet.split())
+                packet, self.buf = self.get_packet(self.buf)
+                continue
+
+            if self.verbose:
+                print(d)
+            assert len(d) == 4, d
+            prev = self.encoder
+            self.milisec = d[0]
+            self.encoder = d[1], d[2]  # more inputs expected
+            if prev is not None:
+                left, right = self.encoder[0] - prev[0], self.encoder[1] - prev[1]
+                self.publish('encoders', [left, right])
+                valid, pose, dist_angle = self.compute_pose(left, right)
+                if valid:
+                    self.pose = pose
+                self.send_pose()
+
+            packet, self.buf = self.get_packet(self.buf)
+
+    def slot_tick(self, timestamp, data):
```

Comment:
So this actually implements speed control?

Reply:
the speed control is planned for next iteration ... now it only drives with `osgar.go` and parameter `--speed` *100 is PWM

Reply:
can we have a more descriptive name here? Maybe
```suggestion
    def slot_pwm(self, timestamp, data):
```

Reply:
well, the input is "tick", i.e. periodically called function. PWM is current workaround to test the platform - soon (in a week?) there should be some PID controller, i.e. pwm name would be wrong

Reply:
The suggested naming is based on what it does. When it does PID, let's rename it. When I see the connection `timer.tick` -> `maria.slot_tick` I have close to no information on what the intention is. Compare that to `timer.tick` -> `maria.slot_pwm` or `maria.slot_pid`.